### PR TITLE
Cosmos change feed support for pull and push model in cosmos encryption project

### PR DESCRIPTION
**Pull request by simplynaveen20**
This PR contains below changes

1. ChangeFeed pull model via queryChangeFeed api on CosmosEncryptionAsyncContainer and CosmosEncryptionContainer, added beta tag as cosmos core sdk haven't done the GA of this api.
2. ChangeFeed push model by creating new ChangeFeedEncryptionProcessorBuilder which takes encryption container in feedContainer api.
3. Did some refactoring of test classes,sharing encryption database/container across tests , moved common code in base class TestSuiteBase

### CODE REVIEW

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/ChangeFeedEncryptionProcessorBuilder.java [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    public ChangeFeedEncryptionProcessorBuilder handleChanges(Consumer<List<JsonNode>> consumer) {
+        this.encryptionConsumer = jsonNodes -> {
+            List<byte[]> byteArrayList = jsonNodes.stream()
+                .map(node -> feedContainer.cosmosSerializerToStream(node))
+                .collect(Collectors.toList());
+            List<Mono<byte[]>> byteArrayMonoList =
+                jsonNodes.stream().map(jsonNode -> feedContainer.decryptResponse((ObjectNode) jsonNode)).collect(Collectors.toList());
+            Flux.concat(byteArrayMonoList).publishOn(Schedulers.boundedElastic()).map(
+                item -> feedContainer.getItemDeserializer().parseFrom(JsonNode.class, item)
+            ).collectList().doOnSuccess(consumer).block();
```
Comment (simplynaveen20) [inline sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/ChangeFeedEncryptionProcessorBuilder.java:93]: We have to do blocking call as change feed processor using synchronous error handling to kill the worker. https://github.com/Azure/azure-sdk-for-java/blob/6fef0a30aeab08cb6a9c3db75a26bfdf2b6b16fc/sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/implementation/DefaultObserver.java#L40

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/CosmosEncryptionAsyncContainer.java [hunk 2]
```diff
@@ -270,6 +279,15 @@ EncryptionProcessor getEncryptionProcessor() {
         }
     }
 
+    @Beta(value = Beta.SinceVersion.V1, warningText =
+        Beta.PREVIEW_SUBJECT_TO_CHANGE_WARNING)
+    public <T> CosmosPagedFlux<T> queryChangeFeed(CosmosChangeFeedRequestOptions options, Class<T> classType) {
+        checkNotNull(options, "Argument 'options' must not be null.");
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/CosmosEncryptionAsyncContainer.java:285]: nice validation from guava.

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/ChangeFeedEncryptionProcessorBuilder.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+import java.util.List;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+
+/**
+ * Helper class to build a encryption supported {@link ChangeFeedProcessor} instance.
+ *
+ */
+public class ChangeFeedEncryptionProcessorBuilder {
```
Comment (milismsft) [inline sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/ChangeFeedEncryptionProcessorBuilder.java:24]: Beta annotations?

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/CosmosEncryptionDatabase.java [hunk 4]
```diff
@@ -105,7 +105,7 @@ public CosmosEncryptionContainer getCosmosEncryptionContainer(CosmosContainer co
      * @param containerId original container id
      * @return container with encryption capabilities
      */
-    public CosmosEncryptionContainer getCosmosEncryptionAsyncContainer(String containerId) {
+    public CosmosEncryptionContainer getCosmosEncryptionContainer(String containerId) {
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/CosmosEncryptionDatabase.java:108]: nice catch

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosChangeFeedRequestOptions.java [hunk 5]
```diff
@@ -432,4 +434,49 @@ public CosmosChangeFeedRequestOptions setThroughputControlGroupName(String throu
...
+
+                @Override
+                public CosmosChangeFeedRequestOptions setHeader(CosmosChangeFeedRequestOptions changeFeedRequestOptions, String name, String value) {
+                    return changeFeedRequestOptions.setHeader(name, value);
+                }
+
+                @Override
+                public Map<String, String> getHeader(CosmosChangeFeedRequestOptions changeFeedRequestOptions) {
+                    return changeFeedRequestOptions.getHeaders();
+                }
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosChangeFeedRequestOptions.java:479]: thank's for following this pattern. this is nice.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java [hunk 6]
```diff
@@ -1688,4 +1689,20 @@ public void enableGlobalThroughputControlGroup(
...
         this.database.getClient().enableThroughputControlGroup(globalControlGroup);
     }
+
+    ///////////////////////////////////////////////////////////////////////////////////////////
+    // the following helper/accessor only helps to access this class outside of this package.//
+    ///////////////////////////////////////////////////////////////////////////////////////////
+
+    static {
+        ImplementationBridgeHelpers.CosmosAsyncContainerHelper.setCosmosAsyncContainerAccessor(
+            new ImplementationBridgeHelpers.CosmosAsyncContainerHelper.CosmosAsyncContainerAccessor() {
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java:1699]: nice

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/ChangeFeedEncryptionProcessorBuilder.java (hunk 3 above)
Comment (simplynaveen20) [reply sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/ChangeFeedEncryptionProcessorBuilder.java:24]: @kushagraThapar  we decided not to add beta to all the classes and methods (which is GA in core project) of encryption as we have not done GA.
So I assume it will be applicable to this class as well. Please share the thoughts

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/ChangeFeedEncryptionProcessorBuilder.java (hunk 1 above)
Comment (moderakh) [reply sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/ChangeFeedEncryptionProcessorBuilder.java:87]: thanks for explanation.

Later this consumer may be invoked on the io thread. Does that cause any problem (thread deadlock, etc)? or the fact that you are using publishOn takes care of that?

@milismsft is there any other workaround?

**Review by moderakh — APPROVED** [review]
LGTM. as this is CFP related please ask @milismsft to review/sign off as well.

Thanks Naveen.

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/ChangeFeedEncryptionProcessorBuilder.java (hunk 1 above)
Comment (simplynaveen20) [reply sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/ChangeFeedEncryptionProcessorBuilder.java:87]: @moderakh  yes for the same concern of block on IO thread i moved the publisher on bounded elastic where we can do the blocking call. This will handled the scenerio

Comment (milismsft) [reply sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/ChangeFeedEncryptionProcessorBuilder.java:87]: A better option is to have the encryption implement its own ChangeFeedObserver class (see https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/ChangeFeedObserver.java). You can also add a property into the ChangeFeedProcessorBuilderImpl to toggle between regular CFP and encrypted CFP when calling the "factory" class to create the respective instances.

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/ChangeFeedEncryptionProcessorBuilder.java [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+    /**
+     * Sets a consumer function which will be called to process changes.
+     *
+     * @param consumer the {@link Consumer} to call for handling the feeds.
+     * @return current Builder.
+     */
+    public ChangeFeedEncryptionProcessorBuilder handleChanges(Consumer<List<JsonNode>> consumer) {
+        this.encryptionConsumer = jsonNodes -> {
+            List<Mono<byte[]>> byteArrayMonoList =
+                jsonNodes.stream().map(jsonNode -> feedContainer.decryptResponse((ObjectNode) jsonNode)).collect(Collectors.toList());
```
Comment (milismsft) [inline sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/ChangeFeedEncryptionProcessorBuilder.java:84]: ObjectNode objectNode = jsonNode.deepCopy();

https://stackoverflow.com/questions/32713109/how-to-convert-jsonnode-to-objectnode

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosChangeFeedRequestOptions.java [hunk 8]
```diff
@@ -432,4 +434,49 @@ public CosmosChangeFeedRequestOptions setThroughputControlGroupName(String throu
...
+    ///////////////////////////////////////////////////////////////////////////////////////////
+    // the following helper/accessor only helps to access this class outside of this package.//
+    ///////////////////////////////////////////////////////////////////////////////////////////
+
+    static {
+        ImplementationBridgeHelpers.CosmosChangeFeedRequestOptionsHelper.setCosmosChangeFeedRequestOptionsAccessor(
+            new ImplementationBridgeHelpers.CosmosChangeFeedRequestOptionsHelper.CosmosChangeFeedRequestOptionsAccessor() {
+
+                @Override
+                public CosmosChangeFeedRequestOptions setHeader(CosmosChangeFeedRequestOptions changeFeedRequestOptions, String name, String value) {
```
Comment (milismsft) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosChangeFeedRequestOptions.java:472]: I don't understand why is this needed? can we get away from it by toggling some boolean in the ChangeFeedProcessorBuilderImpl or something like that depending on encryption or not cases? This is only needed for a set of specific values that need to be added and not some arbitrary ones, right?

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/ChangeFeedEncryptionProcessorBuilder.java (hunk 1 above)
Comment (simplynaveen20) [reply sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/ChangeFeedEncryptionProcessorBuilder.java:87]: @milismsft Encryption have to access the ChangeFeedProcessorBuilderImpl which is implementation class in cosmos project, also ChangeFeedProcessorBuilderImpl will have to handle encryption logic, we want cosmos sdk to have minimal encryption logic .
IMA if we can avoid cosmos dependency on encryption project and treat encryption project as a wrapper as much as possible that would be good.
Coming back to this blocking call, I think ideal solution would have been the async consumer in core sdk , but that would breaking change now.
What concern we have with the current approach, and if we have very strong opinion against current design , i am open to other approach as well.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosChangeFeedRequestOptions.java (hunk 8 above)
Comment (simplynaveen20) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosChangeFeedRequestOptions.java:472]: This is for to pass encryption related header in cf request , this is line with item and query options. This is non public method so we shoudl be fine

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/ChangeFeedEncryptionProcessorBuilder.java (hunk 7 above)
Comment (simplynaveen20) [reply sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/ChangeFeedEncryptionProcessorBuilder.java:84]: In our case all these are object node , we don't need deepcopy.
 We will revisit logic once we work on this work item of aggregated queries https://github.com/Azure/azure-sdk-for-java/issues/23160

Comment (simplynaveen20) [reply sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/ChangeFeedEncryptionProcessorBuilder.java:84]: Discussed offline, added object node check and throw custom exception if it is not

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosChangeFeedRequestOptions.java (hunk 8 above)
Comment (simplynaveen20) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosChangeFeedRequestOptions.java:472]: Discussed offline, this is fine

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/ChangeFeedEncryptionProcessorBuilder.java (hunk 1 above)
Comment (simplynaveen20) [reply sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/ChangeFeedEncryptionProcessorBuilder.java:87]: Discussed offline with @milismsft  we are good here from this PR scope, however created work item to add support in change feed processor to support customer observer https://github.com/Azure/azure-sdk-for-java/issues/23738

File: sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/ChangeFeedEncryptionProcessorBuilder.java (hunk 3 above)
Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/ChangeFeedEncryptionProcessorBuilder.java:24]: @milismsft  - We don't need Beta annotations for this one - since the whole project is in Beta preview only.
Once we decide to GA, then we can add Beta annotations if we need to, but for now, we don't need those. 
To be clear - all the new classes in azure-cosmos project related to encryption should have `@Beta` annotation. 
All new classes in azure-cosmos-encryption project need not to have any `@Beta` annotation.

Comment (simplynaveen20) [reply sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/encryption/ChangeFeedEncryptionProcessorBuilder.java:24]: Thanks @kushagraThapar  for confirming

**Review by milismsft — APPROVED** [review]
